ocaml/testsuite/tests/typing-layouts-or-null/variables.ml

```diff
@@ -259,3 +291,79 @@ end
 [%%expect{|
 module M : sig val f : 'a -> 'a end
 |}]
+
+
+type ('a : value_or_null) dummy
+
+(* This must infer ('a : value) for backwards compatibility. *)
+type t = Packed : 'a dummy -> t
+
+[%%expect{|
+type ('a : value_or_null) dummy
+type t = Packed : 'a dummy -> t
+|}]
+
+(* CR layouts v3.0: annotations here are upper bounds, so
+   due to defaulting we can't set ['a : value_or_null].
+   Fixing this might be hard, see [Note about [new_var_jkind]]. *)
+type t = Packed : ('a : value_or_null) dummy -> t
+[%%expect{|
+type t = Packed : 'a dummy -> t
+|}]
+
+(* CR layouts v3.0: variables on the right side of
+   constraints default to being non-null. This must be the case: the same mechanism is used for . *)
+type 'b constrained = unit constraint 'b = 'a dummy
+
+[%%expect{|
+type 'b constrained = unit constraint 'b = 'a dummy
+|}]
+
+type fails = bool constrained
+
+[%%expect{|
+Line 1, characters 13-17:
+1 | type fails = bool constrained
+                 ^^^^
+Error: This type bool should be an instance of type 'a dummy
+|}]
+
+type fails = (t_value_or_null dummy) constrained
+
+[%%expect{|
+Line 1, characters 14-35:
+1 | type fails = (t_value_or_null dummy) constrained
+                  ^^^^^^^^^^^^^^^^^^^^^
+Error: This type t_value_or_null dummy should be an instance of type 'a dummy
+       The layout of t_value_or_null is value_or_null, because
+         of the definition of t_value_or_null at line 1, characters 0-36.
+       But the layout of t_value_or_null must be a sublayout of value, because
+         of the definition of constrained at line 1, characters 0-51.
+|}]
+
+type succeeds = (int dummy) constrained
+
+[%%expect{|
+type succeeds = int dummy constrained
+|}]
+
+(* CR layouts v3.0: we can't set a variable on the right side
+   of the constraint to be [or_null]. This might be unfixable. *)
```

Comment:
I'm worried about this, too. We probably need a

```
type 'a 'b 'c. 'c constrained' = bool constraint 'c = ('a, 'b) result
```

syntax or something, where we can put definitive kind annotations on the binders at the beginning.

But let's wait to write this syntax until we need it. Constraints are not common.

Reply:
Why add new syntax instead of changing the behavior of ascriptions to the intuitive one? Needs less changes to support in public release too.

Reply:
Mostly because I don't know how to do this. Type variables can be born with any kind. But once they're born, they can generally only be constrained to be less than some other kind. I can imagine some pre-pass that looks for ascriptions and has the type variable born at the right kind, but then what do we do with `val f : ('a : any) -> ('a : value)`? If `'a : value`, then `'a : any` is just fine.